parallel backend for `foreach` using `spark_apply`

yitao-li opened the pull request:
This PR creates a `foreach` parallel backend based on `spark_apply`, as described in  https://github.com/sparklyr/sparklyr/issues/2130

yitao-li commented:
just fixed failing tests from previous revision

yitao-li commented on R/do_spark.R:
```diff
@@ -0,0 +1,126 @@
+    spark_conn <- data$spark_conn
+    spark_apply_args <- data$spark_apply_args
+
+    it <- iterators::iter(obj)
```
Most likely not, because the accumulator functionality in `foreach` is implemented in term of iterator:

```
> makeAccum                                                                                                                                                                                                 function (it)                                                                                                                                                                                                      
{                                                                                                                                                                                                                  
    function(results, tags) {                                                                                                                                                                                      
        if (identical(it$error.handling, "stop") && !is.null(it$state$errorValue)) 
            return(invisible(NULL))
        for (i in seq(along.with = tags)) {
            if (it$verbose) 
                cat(sprintf("got results for task %d\n", tags[i]))
            accumulate(it, results[[i]], tags[i])
        }
    }
}
```

However, because `iterators` is definitely a dependency of the `foreach` package already, I think all we have to do is to list `foreach` as a suggested dependency for `sparklyr` so that `iterators` package will not be automatically installed when installing `sparklyr`.

yitao-li commented:
closing & re-opening PR to restart travis build

javierluraschi commented:
@yitao-li there might be an invalid character somewhere in the file, see test output:

```
Installing package into ‘/home/travis/R/Library’
(as ‘lib’ is unspecified)
* installing *source* package ‘sparklyr’ ...
** R
Error in parse(outFile) : 
  /tmp/RtmpMSKKFe/R.INSTALL36f052376d13/sparklyr/R/do_spark.R:5:3: unexpected symbol
4:   # options from the last successful call to registerDoSpark
5:   options
     ^
ERROR: unable to collate and parse R files for package ‘sparklyr’
* removing ‘/home/travis/R/Library/sparklyr’
```

In addition, you might want to rebase to your now merged PR: https://github.com/sparklyr/sparklyr/pull/2261

yitao-li commented:
> @yitao-li there might be an invalid character somewhere in the file, see test output:
> 
> ```
> Installing package into ‘/home/travis/R/Library’
> (as ‘lib’ is unspecified)
> * installing *source* package ‘sparklyr’ ...
> ** R
> Error in parse(outFile) : 
>   /tmp/RtmpMSKKFe/R.INSTALL36f052376d13/sparklyr/R/do_spark.R:5:3: unexpected symbol
> 4:   # options from the last successful call to registerDoSpark
> 5:   options
>      ^
> ERROR: unable to collate and parse R files for package ‘sparklyr’
> * removing ‘/home/travis/R/Library/sparklyr’
> ```
> 
> In addition, you might want to rebase to your now merged PR: #2261

@javierluraschi  yes indeed... just rebased, fixed syntax error, and trying to re-run CI jobs now.

yitao-li commented:
hmm just realized something went wrong with `.globals` -- fixing it now

javierluraschi commented:
You probably already noticed this but yeah, this is the extract from the failed tests:

```
* checking dependencies in R code ... WARNING
'::' or ':::' import not declared from: ‘iterators’
* checking S3 generic/method consistency ... OK
* checking replacement functions ... OK
* checking foreign function calls ... OK
* checking R code for possible problems ... NOTE
registerDoSpark : .doSpark: no visible global function definition for
  ‘makeAccum’
registerDoSpark : .doSpark: no visible global function definition for
  ‘getErrorValue’
registerDoSpark : .doSpark: no visible global function definition for
  ‘getErrorIndex’
registerDoSpark : .doSpark: no visible global function definition for
  ‘getResult’
registerDoSpark: no visible global function definition for ‘setDoPar’
Undefined global functions or variables:
  getErrorIndex getErrorValue getResult makeAccum setDoPar
* checking Rd files ... OK
```

yitao-li commented:
Finally fixed all test failures (except for the "failed to launch livy session" one which appears to be unrelated)

yitao-li commented on R/do_spark.R:
```diff
@@ -0,0 +1,126 @@
+  .doSpark <- function(obj, expr, envir, data) {
+    # internal function to compile an expression if possible
+    .compile <- function(expr, ...) {
+      if (getRversion() < "2.13.0" || isTRUE(.options$nocompile))
```
compile/nocompile seems to make little difference for simple expressions (e.g., the ones in unit tests):

compiled expressions:

Context: do-spark                                                                                                                                                                    
doSpark works for simple loop: 39.9492995738983                                                                                                                                      
doSpark works for simple loop with combine function: 24.6455733776093
doSpark works for simple loop of matrices: 22.6680960655212
doSpark works for loop with arbitrary R objects: 8.60413360595703
doSpark works for loop with arbitrary R objects with combine function: 8.78711891174316
doSpark works for loop with arbitrary R objects with multicombine: 41.8657147884369


--- Performance Summary  ----

  context      time
 do-spark       147


non-compiled expressions:

Context: do-spark                                                                                                                                                                    
doSpark works for simple loop: 44.2782394886017                                                                                                                                      
doSpark works for simple loop with combine function: 22.7579934597015
doSpark works for simple loop of matrices: 21.8420250415802
doSpark works for loop with arbitrary R objects: 8.14156150817871
doSpark works for loop with arbitrary R objects with combine function: 8.08031034469604
doSpark works for loop with arbitrary R objects with multicombine: 39.6312344074249


--- Performance Summary  ----

  context      time
 do-spark       145

Total: 145s

yitao-li replied:
but it probably makes more difference for non-trivial expressions (although I haven't managed to find anything sufficiently complicated that would show noticeable difference yet...)
